File: src/main/scala/com/zotoh/frwk/io/XData.scala
/*??
 * COPYRIGHT (C) 2012 CHERIMOIA LLC. ALL RIGHTS RESERVED.
 *
 * THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE,
 * VERSION 2.0 (THE "LICENSE").
 *
 * THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
 * AND LIMITATIONS UNDER THE LICENSE.
 *
 * You should have received a copy of the Apache License
 * along with this distribution; if not, you may obtain a copy of the
 * License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 ??*/

package com.zotoh.frwk
package io

import java.io.{ByteArrayOutputStream=>ByteArrayOS,File,InputStream}
import org.apache.commons.lang3.{StringUtils=>STU}
import org.apache.commons.io.{FileUtils=>FUT}
import org.slf4j._
//import com.zotoh.frwk.util.FileUtils
import com.zotoh.frwk.util.CoreUtils._
import com.zotoh.frwk.util.StrUtils._
import com.zotoh.frwk.io.IOUtils._


/**
 * Wrapper structure to abstract a piece of data which can be a file
 * or a memory byte[].  If the data is byte[], it will also be
 * compressed if a certain threshold is exceeded.
 *
 * @author kenl
 *
 */

object XData {

  private val _log= LoggerFactory.getLogger(classOf[XData])
  private var _wd = tmpDir()

  /**
   * Get the current shared working directory.
   *
   * @return working dir.
   */
  def workDir = _wd

  /**
   * Working directory is where all the temporary/sharable files are created.
   *
   * @param fpDir a directory.
   */
  def workDir_=(fpDir:File) {
    block { () =>
      if (fpDir != null && fpDir.isDirectory &&
        fpDir.canRead && fpDir.canWrite) {
        _wd=fpDir
      }
    }
  }

}

@SerialVersionUID(-8637175588593032279L)
class XData() extends Serializable {

  import XData._
  def tlog() = _log

  private val CMPZ_THREADHOLD=1024*1024*4 // 4meg
  private var _cmpz=false
  private var _cls=true

  private var _encoding="UTF-8"
  private var _fp=""

  private var _bin:Array[Byte]= null
  private var _binSize= 0L

  /**
   *
   */
  def javaBytes() = bytes().getOrElse( Array[Byte]() )

  /**
   *
   */
  def this(p:AnyRef) {
    this()
    resetMsgContent(p)
  }

  /**
   * @param enc
   */
  def encoding_=(enc:String) { _encoding=enc }

  /**
   * @return
   */
  def encoding = _encoding

  /**
   *
   */
  def isZiped =  _cmpz

  /**
   * Control the internal file.
   *
   * @param del true to delete, false ignore.
   */
  def setDeleteFile(del:Boolean): this.type = { _cls= del; this }

  /**
   * Tests if the file is to be deleted.
   *
   * @return
   */
  def isDeleteFile() = _cls

  /**
   * Clean up.
   */
  def destroy() {
    if (! STU.isEmpty(_fp) && isDeleteFile ) {
      FUT.deleteQuietly( new File(_fp))
    }
    reset()
  }

  /**
   * Tests if the internal data is a file.
   *
   * @return
   */
  def isDiskFile() = ! STU.isEmpty( _fp)


  /**
   * @param obj
   */
  def resetMsgContent(obj:Any): XData = resetMsgContent(obj, true)


  /**
   * @param obj
   * @param delIfFile
   */
  def resetMsgContent(obj:Any, delIfFile:Boolean): this.type = {
    destroy()
    obj match {
      case fa:Array[File] => if (fa.length > 0) _fp= niceFPath( fa(0))
      case baos: ByteArrayOS => maybeCmpz( baos.toByteArray)
      case bits: Array[Byte] => maybeCmpz( bits)
      case f:File => _fp= niceFPath( f)
      case s:String => maybeCmpz( s.getBytes(_encoding))
      case _ =>
    }
    setDeleteFile(delIfFile)
  }

  /**
   * Get the internal data.
   *
   * @return
   */
  def content(): Option[Any] = {
    val a:Any = if (isDiskFile)  {
      new File( _fp)
    }
    else if ( _bin==null) {
      null
    }
    else if (! _cmpz) {
      _bin
    } else {
      gunzip( _bin)
    }

    if (a==null) None else Some(a)
  }

  /**
   * @return
   */
  def hasContent() = {
    if (isDiskFile) true else if ( _bin !=null) true else false
  }

  /**
   * Get the data as bytes.  If it's a file-ref, the entire file content will be read
   * in as byte[].
   *
   * @return
   * @throws IOException
   */
  def bytes(): Option[Array[Byte]] = {
    val b= if (isDiskFile) read(new File( _fp)) else binData()
    if (b==null) None else Some(b)
  }


  /**
   * Get the file path if it is a file-ref.
   *
   * @return the file-path, or null.
   */
  def fileRef() =  if ( STU.isEmpty(_fp)) None else Some(new File( _fp))


  /**
   * Get the file path if it is a file-ref.
   *
   * @return the file-path, or "".
   */
  def filePath() =  if ( STU.isEmpty(_fp)) "" else _fp


  /**
   * Get the internal data if it is in memory byte[].
   *
   * @return
   */
  def binData() = {
    if (_bin==null) null else ( if (!_cmpz) _bin else gunzip( _bin) )
  }

  /**
   * Get the size of the internal data (no. of bytes).
   *
   * @return
   */
  def size() = {
    if ( isDiskFile) new File(_fp).length() else _binSize
  }


  override def finalize() {
    destroy()
  }

  override def toString() = {
    new String ( javaBytes(), _encoding )
  }

  /**
   *
   */
  def stream(): InputStream = {
    if (isDiskFile) { new XStream(new File( _fp)) }
    else if ( _bin==null) { asStream(null) }
    else {
      asStream( if (_cmpz) gunzip(_bin) else _bin )
    }
  }

  private def maybeCmpz(bits:Array[Byte]) {
    _binSize= if (bits==null) 0L else bits.length
    _cmpz=false
    _bin=bits
    if (_binSize > CMPZ_THREADHOLD) {
      _cmpz=true
      _bin= gzip(bits)
    }
  }

  private def reset() {
    _cmpz=false
    _cls=true
    _bin=null
    _fp=null
    _binSize=0L
  }

}

File: src/main/java/com/zotoh/frwk/io/XData.java
/*??
 * COPYRIGHT (C) 2012 CHERIMOIA LLC. ALL RIGHTS RESERVED.
 *
 * THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
 * MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE,
 * VERSION 2.0 (THE "LICENSE").
 *
 * THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL,
 * BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
 *
 * SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
 * AND LIMITATIONS UNDER THE LICENSE.
 *
 * You should have received a copy of the Apache License
 * along with this distribution; if not, you may obtain a copy of the
 * License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 ??*/

package com.zotoh.frwk.io;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import org.apache.commons.lang3.StringUtils;
import org.apache.commons.io.FileUtils;


/**
 * Wrapper structure to abstract a piece of data which can be a file
 * or a memory byte[].  If the data is byte[], it will also be
 * compressed if a certain threshold is exceeded.
 *
 * @author kenl
 *
 */
public class XData implements java.io.Serializable {

  private static final long serialVersionUID = 3109739389414555705L;
  private int CMPZ_THREADHOLD=1024*1024*4; // 4meg
  private boolean _cmpz=false;
  private boolean _cls=true;

  private String _encoding="UTF-8";
  private String _fp="";

  private long _binSize= 0L;
  private byte[] _bin;

  public byte[] javaBytes() {
    byte[] rc= bytes();
    if (rc == null) { return new byte[0]; } else { return rc; }
  }

  public XData( Object p) throws IOException {
    resetContent(p);
  }
  
  public XData() {
  }

  public void setEncoding(String enc) { _encoding=enc; }

  public String getEncoding() { return _encoding; }

  public boolean isZiped() { return  _cmpz; }

  /**
   * Control the internal file.
   *
   * @param del true to delete, false ignore.
   */
  public XData setDeleteFile(boolean del) {
    _cls= del;
    return this; 
  }

  /**
   * Tests if the file is to be deleted.
   *
   * @return
   */
  public boolean isDeleteFile() { return _cls; }

  /**
   * Clean up.
   */
  public void destroy() {
    if (! StringUtils.isEmpty(_fp) && isDeleteFile() ) {
      FileUtils.deleteQuietly( new File(_fp));
    }
    reset();
  }

  /**
   * Tests if the internal data is a file.
   *
   * @return
   */
  public boolean isDiskFile() { 
    return ! StringUtils.isEmpty( _fp);
  }

  public XData resetContent(Object obj) throws IOException { 
    return resetContent(obj, true);
  }

  public XData resetContent(Object obj, boolean delIfFile) throws IOException {
    destroy();
    if (obj instanceof File[]) {
      File[] fa= (File[]) obj;
      if (fa.length > 0) {
        _fp= fa[0].getCanonicalPath();
      }
    }
    else
    if (obj instanceof ByteArrayOutputStream) {
      ByteArrayOutputStream baos = (ByteArrayOutputStream ) obj;
      maybeCmpz( baos.toByteArray());
    }
    else
    if (obj instanceof byte[]) {
      maybeCmpz( (byte[]) obj);
    }
    else
    if (obj instanceof File) {
      File f= (File) obj;
      _fp= f.getCanonicalPath();
    }
    else
    if (obj instanceof String) {
      String s= (String) obj;
      maybeCmpz( s.getBytes(_encoding));
    }
    else if (obj != null) {
      throw new IOException("Unsupported data type: " + obj.getClass());
    }
    setDeleteFile(delIfFile);
    return this;
  }

  /**
   * Get the internal data.
   *
   * @return
   */
  public Object content() {
    if (isDiskFile() )  {
      return new File( _fp);
    }
    if ( _bin==null) {
      return null;
    }
    if (! _cmpz) {
      return _bin;
    } else {
      try { return IOUtils.gunzip( _bin); } catch (Throwable t) { return null; }
    }
  }

  public boolean hasContent() {
    if (isDiskFile()) { return true; } 
    else 
    if ( _bin !=null) {
      return true;
    } else {
      return false;
    }
  }

  /**
   * Get the data as bytes.  If it's a file-ref, the entire file content will be read
   * in as byte[].
   *
   * @return
   * @throws IOException
   */
  public byte[] bytes() {
    if (isDiskFile()) {
      try { return FileUtils.readFileToByteArray(new File(_fp)); } catch (Throwable t) { return null; }
    } else {
      return binData();
    }
  }

  /**
   * Get the file path if it is a file-ref.
   *
   * @return the file-path, or null.
   */
  public File fileRef() {
    if ( StringUtils.isEmpty(_fp)) { return null; } else { return new File( _fp); }
  }

  /**
   * Get the file path if it is a file-ref.
   *
   * @return the file-path, or "".
   */
  public String filePath() {
    if ( StringUtils.isEmpty(_fp)) { return ""; } else { return _fp; }
  }


  /**
   * Get the internal data if it is in memory byte[].
   *
   * @return
   */
  public byte[] binData(){
    if (_bin==null) { return null; } else {
      if (!_cmpz) { return _bin; } else { try { return IOUtils.gunzip( _bin); } catch (Throwable t) { return null; } }
    }
  }

  public long size() {
    if ( isDiskFile() ) { return new File(_fp).length(); } else { return _binSize; }
  }

  public void finalize() {
    destroy();
  }

  public String stringify() {
    try { return new String( javaBytes(), _encoding ); } catch (Throwable t) { return null; }
  }

  public InputStream stream() throws IOException {
    if (isDiskFile()) { return new XStream(new File( _fp)); }
    else if ( _bin==null) { return null; }
    else {
      byte[] bits= _bin;
      if (_cmpz) { bits= IOUtils.gunzip(_bin); }
      return new ByteArrayInputStream( bits);
    }
  }

  private void maybeCmpz(byte[] bits) throws IOException {
    if (bits==null) { _binSize= 0L; } else { _binSize= bits.length; }
    _cmpz=false;
    _bin=bits;
    if (_binSize > CMPZ_THREADHOLD) {
      _cmpz=true;
      _bin= IOUtils.gzip(bits);
    }
  }

  private void reset() {
    _cmpz=false;
    _cls=true;
    _bin=null;
    _fp=null;
    _binSize=0L;
  }

}

